segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,385 @@
+                    !(attrs.permissions().contains(OWNER_WRITE)), false,
+                    new ImmutableDate(attrs.creationTime().toMillis()));
+
+            return information;
+        } catch (IOException e) {
+            throw new CompletionException(e);
+        }
+    }
+
+    private boolean syncExists(String streamSegmentName, Duration timeout) {
+        boolean exists = Files.exists(Paths.get(config.getNfsRoot(), streamSegmentName));
+        return exists;
+    }
+
+
+    private SegmentProperties syncCreate(String streamSegmentName, Duration timeout) {
+        log.info("Creating Segment {}", streamSegmentName);
+        try {
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as rw-r--r-- 644
+            perms.add(PosixFilePermission.OWNER_WRITE);
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+            Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+            Files.createDirectories(path.getParent());
+            Files.createFile(path, fileAttributes);
+            log.info("Created Segment {}", streamSegmentName);
+            return this.getStreamSegmentInfo(streamSegmentName, timeout).get();
+        } catch (Exception e) {
+            log.info("Exception {} while creating a segment {}", e, streamSegmentName);
+            if (e instanceof FileAlreadyExistsException) {
+                throw new CompletionException(new StreamSegmentExistsException(streamSegmentName, e));
+            } else {
+                throw new CompletionException(e);
+            }
+        }
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length, Duration timeout) {
+        log.trace("Writing {} to segment {} at offset {}", length, handle.getSegmentName(), offset);
+        Path path = Paths.get(config.getNfsRoot(), handle.getSegmentName());
+
+        if (handle.isReadOnly()) {
+            log.info("Write called on a readonly handle of segment {}", handle.getSegmentName());
+            throw new CompletionException(new IllegalArgumentException());
+        }
+
+        if (!Files.exists(path)) {
+            throw new CompletionException(new StreamSegmentNotExistsException(handle.getSegmentName(), null));
+        }
+
+        try {
+            if (!isWritableFile(path)) {
+                throw new CompletionException(new StreamSegmentSealedException(handle.getSegmentName()));
+            }
+
+            long fileSize = path.toFile().length();
+            if (fileSize < offset) {
+                throw new CompletionException(new BadOffsetException(handle.getSegmentName(), fileSize, offset));
```

Comment:
Is this offset check defensive or is there an expected situation in which we hit this?